Cleanup find_symbols API

tautschnig opened the pull request:
The various functions declared in `find_symbols.h` might sound alike, but actually have different semantics. I wasn't really sure what to use anymore, so I went for a cleanup instead. Requires #6725 and #6726. Please review commit-by-commit. 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Make has_symbol configurable and deprecate the existing has_symbol

The existing code did not permit configuring what kinds of symbols are
to be found. It had a single user within the code base, and even that
one was not a good use of it for it constructed a single-element set.

The new implementation does not yet make use of potential efficiency
advantages of "has a symbol" over "find the existing symbols," but such
improvements will be done based on further cleanup of find_symbols code.

tautschnig committed:
Deprecate find_symbols_or_nexts

Make the two Boolean parameters of find_symbols(expr, set, bool, bool)
optional so that each existing users can safely use that function
(or another, even more suitable, function) instead.

tautschnig committed:
Rename kindt to symbol_kindt for clarity

This is just a change in the enum tag.

tautschnig committed:
find_symbols kindt: clarify enum entries

"F_BOTH" is not a good fit when there are three (and not two) other
entries. While at it, also add support to selectively choose current or
next expression symbols.

tautschnig committed:
find_symbols: support restriction to non-bound symbols

Adds a new symbol kind to distinguish bound from unbound ones. Add
documentation to all non-deprecated functions and clarify whether or not
bound symbols are included.

kroening commented:
This function is trying to do too much.

1. Type-symbols and 'normal ones' are really very different, and the use-cases won't overlap. I'd make these separate functions.
2. I don't think next_symbol is used anywhere in this repo. It can go away.
3. This leaves bound vs. non-bound symbols. How about passing that as a category to a functor?

tautschnig commented:
> This function is trying to do too much.
> 
>     1. Type-symbols and 'normal ones' are really very different, and the use-cases won't overlap. I'd make these separate functions.

That makes a lot of sense as far as the user-facing API is concerned (which already has different functions), but I'm not sure the underlying implementation would be very different? Expressions can be contained in types and types in expressions, so the iteration will be rather similar?

> 
>     2. I don't think next_symbol is used anywhere in this repo. It can go away.
> 

The `precondition` and `postcondition` code presumably still supports it, but likely no one cares. I'll clean this up.

>     3. This leaves bound vs. non-bound symbols. How about passing that as a category to a functor?

With the above changes this might end up as nothing more than a Boolean? I'll give it a try.

kroening commented:
> That makes a lot of sense as far as the user-facing API is concerned (which already has different functions), but I'm not sure the underlying implementation would be very different? Expressions can be contained in types and types in expressions, so the iteration will be rather similar?

I can certainly see a case for offering separate, easy functions as part of the user-facing API, which then call one (complex) function internally to avoid code duplication.

> The `precondition` and `postcondition` code presumably still supports it, but likely no one cares. I'll clean this up.

`next_symbol` is not to be confused with `__CPROVER_old` -- I don't believe pre/postconditions ever used `next_symbol`.

> With the above changes this might end up as nothing more than a Boolean? I'll give it a try.

Super!

tautschnig committed:
Remove support for ID_next_symbol

This is never used in CBMC's code base.

tautschnig committed:
Make find_symbols implementations uniform

Each function used a different approach, with different potential
omissions of symbols. Instead, use a single way to traverse expressions
(and types), and configure the operation done via a parameter. All other
(non-deprecated) functions are now just thin wrappers.

tautschnig committed:
Restrict has_symbol to symbol expressions and rename it to has_symbol_expr

With the removal of support for ID_next_symbol all uses of `has_symbol`
searched for symbol expressions only.

tautschnig commented:
@kroening I hope that the state at the end of commit series in this PR matches your expectations. The individual commits should be meaningful, but there is some back&forth, so it's probably best to either checkout the branch or just review all code changes at once.

kroening commented on src/util/find_symbols.h:
```diff
@@ -10,55 +10,71 @@ Author: Daniel Kroening, kroening@kroening.com
+  return identifiers;
+}
 
+DEPRECATED(SINCE(2022, 3, 14, "use has_symbol(exprt, irep_idt, symbol_kindt)"))
```
the symbol_kindt is likely a remnant

kroening commented:
1. Comments: you may wish to call the non-bound variables the 'free variables'.
2. Can you add an explicit comment whether or not bound variables are considered?
3. Is there a use-case for finding both free and the bound variables in the codebase?

tautschnig commented:
> 1. Comments: you may wish to call the non-bound variables the 'free variables'.

Thank you, done.

> 2. Can you add an explicit comment whether or not bound variables are considered?

Done.

> 3. Is there a use-case for finding both free and the bound variables in the codebase?

I cannot rule out that such a case exists, but I'm not currently aware of one. There may well be places that currently feature a local lookalike of `find_symbols`.

kroening commented:
> > 5\. Is there a use-case for finding both free and the bound variables in the codebase?
> 
> I cannot rule out that such a case exists, but I'm not currently aware of one. There may well be places that currently feature a local lookalike of `find_symbols`.

Let's go with the PR as is for now. In the long term, it makes sense to change the default semantics of the `find_symbol` functions to exclude the bound symbols. I strongly suspect that this is what all users want.
